subsys/net/lib/async_socket/async_socket.c

```diff
@@ -0,0 +1,380 @@
+/*
+ * Copyright (c) 2018, Texas Instruments Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_ASYNC_SOCKETS)
+#define SYS_LOG_DOMAIN "net/async_sock"
+#define NET_LOG_ENABLED 1
+#endif
+
+/* Zephyr headers */
+#include <kernel.h>
+#include <init.h>
+#include <net/async_socket.h>
+
+/* Convenience macro to check for invalid socket fd */
+#define INVALID_SOCK (-1)
+
+/* Add one for signal_socket to unblock poll() on demand: */
+#define MAX_RCV_CALLBACKS (CONFIG_NET_SOCKETS_POLL_MAX + 1)
+
+#define LOOPBACK_ADDR "127.0.0.1"
+#define DISCARD_PORT 9
+
+#define SIGNAL_MSG "POLL"
+#define SIGNAL_MSG_SIZE 4
+
+#define ASYNC_SOCK_TASK_STACKSIZE 1024
+#define ASYNC_SOCK_TASK_PRIORITY K_HIGHEST_THREAD_PRIO
+
+/*Socket to callback map, for use by async_recv() and async_sock_server() */
+struct rcv_callbacks {
+	/* Our socket id */
+	int sock;
+
+	/* Buffer for receive and its maximum length */
+	void *buf;
+	size_t max_len;
+
+	/* Callback to be triggered when socket receives data */
+	async_recv_cb_t cb;
+
+	/* User data to pass back to the callback */
+	void *cb_data;
+};
+static struct rcv_callbacks rcv_callback[MAX_RCV_CALLBACKS] = { 0, };
+static int num_rcv_cbs;
+
+/* Special loopback socket used to unblock the poll() API: */
+static int signal_sock;
+static struct sockaddr_in loopback_addr;
+
+/* Dedicated buffer for loopback message: */
+static u8_t server_rcv_buf[SIGNAL_MSG_SIZE];
+
+/* Loopback socket's bind address: */
+static struct sockaddr_in bind_addr = {
+	.sin_family = AF_INET,
+	.sin_port = htons(DISCARD_PORT),
+	.sin_addr = {
+		.s_addr = htonl(INADDR_ANY),
+	},
+};
+
+/* Async socket server thread: */
+static K_THREAD_STACK_DEFINE(async_sock_task_stack, ASYNC_SOCK_TASK_STACKSIZE);
+static struct k_thread async_sock_task_data;
+
+/* Helper function, to restart poll() loop: */
+static int async_server_restart(void)
```

Comment:
If there weren't a thread, there wouldn't be needed such a complication: polling vs application callbacks were serialized in runtime, i.e. first event loop's poll() runs, then callbacks are served which may update poll structures, then event loop gets control again, which makes needed pass on (updated) poll structures and calls poll() again. Won't fit existing MQTT API (which is API of the broken implementation, to remind again). Much more efficient. (Also potentially more bug-free.)